Userland/Libraries/LibJS/Runtime/RegExpPrototype.cpp

```diff
@@ -281,4 +282,119 @@ JS_DEFINE_NATIVE_FUNCTION(RegExpPrototype::symbol_match)
     return vm.call(*exec, rx, js_string(vm, s));
 }
 
+JS_DEFINE_NATIVE_FUNCTION(RegExpPrototype::symbol_replace)
+{
+    auto string_value = vm.argument(0);
+    auto replace_value = vm.argument(1);
+
+    // https://tc39.es/ecma262/#sec-regexp.prototype-@@replace
+    auto rx = regexp_object_from(vm, global_object);
+    if (!rx)
+        return {};
+    auto string = string_value.to_string(global_object);
+    if (vm.exception())
+        return {};
+
+    auto global_value = rx->get(vm.names.global).value_or(js_undefined());
+    if (vm.exception())
+        return {};
+
+    bool global = global_value.to_boolean();
+    if (global)
+        rx->regex().start_offset = 0;
+
+    // FIXME: Implement and use RegExpExec - https://tc39.es/ecma262/#sec-regexpexec
+    auto* exec = get_method(global_object, rx, vm.names.exec);
+    if (!exec)
+        return {};
+
+    Vector<Value> results;
+
+    while (true) {
+        auto result = vm.call(*exec, rx, string_value);
+        if (result.is_null()) {
+            break;
+        }
+
+        results.append(move(result));
+        if (!global) {
+            break;
+        }
+
+        String match_str = results.last().as_array().get(0).to_string(global_object);
+        if (vm.exception())
+            return {};
+        if (match_str.is_empty()) {
+            // FIXME: Implement AdvanceStringIndex to take Unicode code points into account - https://tc39.es/ecma262/#sec-advancestringindex
+            //        Once implemented, step (8a) of the @@replace algorithm must also be implemented.
+            rx->regex().start_offset += 1;
+        }
+    }
+
+    String accumulated_result;
+    size_t next_source_position = 0;
+
+    for (auto& result : results) {
+        const auto& result_array = result.as_array();
+
+        size_t result_length = length_of_array_like(global_object, result_array);
+        size_t n_captures = result_length == 0 ? 0 : result_length - 1;
+
+        Value matched = result_array.get(0);
+
+        double position = result_array.get(vm.names.index).to_integer_or_infinity(global_object);
+        position = clamp(position, static_cast<double>(0), static_cast<double>(string.length()));
+
+        Vector<Value> captures;
+        for (size_t n = 1; n <= n_captures; ++n) {
+            captures.append(result_array.get(n));
```

Comment:
Missing exception check

Reply:
Yeah I wasn't sure about the `ToString` here. We have to store the `captures` list as a list of `Value`, so we'd basically be converting from a `Value` to a `String` then immediately back to a `Value`. Seemed wasteful since all we do with this information currently is pass it to the function callback. Would just adding an `is_string` check suffice?

Reply:
Calling `to_string()` on a `Value` can also side effects, so we can't just skip it - it is what it is 🤷
We get an `AK::String` and put that into a `JS::PrimitiveString` which is then wrapped in a `JS::Value` (there's no conversion for that) - it's probably not as wasteful as it may seem.

Reply:
:+1: got it. Updated with that step added.